python/lsst/ts/mtdomegui/model.py

```diff
@@ -62,11 +125,605 @@ def __init__(
+                    COMMANDS_REPLIED_PERIOD,
+                )
+            )
+        )
+
+        self.periodic_tasks.append(
+            asyncio.create_task(
+                self.one_periodic_task(
+                    self.mtdome_com.process_command_queue,
+                    _COMMAND_QUEUE_PERIOD,
+                )
+            )
+        )
+
+    async def cancel_periodic_tasks(self) -> None:
+        """Cancel all periodic tasks."""
+
+        while self.periodic_tasks:
+            periodic_task = self.periodic_tasks.pop()
+            periodic_task.cancel()
+            await periodic_task
+
+    async def one_periodic_task(self, method: typing.Callable, interval: float) -> None:
+        """Run one method forever at the specified interval.
+
+        Parameters
+        ----------
+        method : `typing.Callable`
+            The coroutine to run periodically.
+        interval : `float`
+            The interval at which to run the status method in second.
+        """
+
+        self.log.debug(f"Starting periodic task {method=} with {interval=}")
+
+        try:
+            while True:
+                await method()
+                await asyncio.sleep(interval)
+
+        except asyncio.CancelledError:
+            # Ignore because the task was canceled on purpose.
+            pass
+
+        except Exception:
+            self.log.exception(f"one_periodic_task({method}) has stopped.")
+
+    async def disconnect(self) -> None:
+        """Disconnect from the controller, if connected, and stop the
+        mock controller, if running.
+
+        This function is safe to call even if the controller is not connected.
+        """
+
+        self.log.info("disconnect.")
+
+        # Stop all periodic tasks, including polling for the status of the
+        # lower level components.
+        await self.cancel_periodic_tasks()
+
+        # Disconnect from the controller
+        if self.is_connected():
+
+            # Workaround the mypy check
+            assert self.mtdome_com is not None
+
+            await self.mtdome_com.disconnect()
+            self.mtdome_com = None
+
+    def assert_is_connected(self) -> None:
+        """Assert the connection is established.
+
+        Raises
+        ------
+        `RuntimeError`
+            If the connection is not established.
+        """
+
+        if not self.is_connected():
+            raise RuntimeError("The connection is not established.")
+
+    def is_connected(self) -> bool:
+        """Is the controller connected or not.
+
+        Returns
+        -------
+        `bool`
+            True if the controller is connected. False otherwise.
+        """
+
+        return (
+            self.mtdome_com is not None
+            and self.mtdome_com.client is not None
+            and self.mtdome_com.client.connected
+        )
+
+    async def request_llc_status_and_report(self, llc_name: LlcName) -> None:
```

Comment:
This is an exact copy of the method in ts_mtdome. We should extract the one from ts_mtdome and put it in ts_mtdomecom.

Reply:
Done in the commit: [17d71c3](https://github.com/lsst-ts/ts_mtdomegui/pull/4/commits/17d71c31a9d25a477c45cf52d78acfc78b39afce). Thanks!